api/group_creation/group_manager.py

```diff
@@ -0,0 +1,150 @@
+from api.models.choices.log_event_type import (
+    GroupLogEventType,
+    StudentLogEventType,
+    TeacherLogEventType,
+)
+from api.models.choices.status import GroupStatus, TeacherStatus
+from api.models.language_and_level import LanguageAndLevel
+from api.processors.auxil.log_event_creator import GroupLogEventCreator
+
+
+class GroupManager:
+    @dataclass
+    class GroupSizeRestriction:
+        min_number: int
+        max_number: int
+
+    @dataclass
+    class GroupCandidate:
+        language_and_level: LanguageAndLevel
+        communication_language_mode: CommunicationLanguageMode
+        age_range: AgeRange
+        teacher: Teacher
+        students: list[Student]
+        monday: datetime | None = None
+        tuesday: datetime | None = None
+        wednesday: datetime | None = None
+        thursday: datetime | None = None
+        friday: datetime | None = None
+        saturday: datetime | None = None
+        sunday: datetime | None = None
+
+        def is_acceptable(self) -> bool:
+            # Checks group size is appropriate for age range
+            restrictions = GroupManager._get_allowed_group_size(self.age_range)
+            students_num = len(self.students)
+            if restrictions.min_number > students_num:
+                return False
+            if restrictions.max_number < students_num:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {students_num} > {restrictions.max_number}"
+                )
+            return True
+
+    @staticmethod
+    def get_available_teachers() -> Iterable[Teacher]:
+        return Teacher.objects.filter(
+            status__in=(
+                TeacherStatus.AWAITING_OFFER,
+                TeacherStatus.TEACHING_ACCEPTING_MORE,
+            )
+        )
+
+    @staticmethod
+    def create_group(teacher_id: int) -> Group | None:
+        for group_candidate in GroupManager._iterate_group_candidates(teacher_id):
+            if group_candidate.is_acceptable():
+                group = Group(
+                    language_and_level=group_candidate.language_and_level,
+                    communication_language_mode=group_candidate.communication_language_mode,
+                    lesson_duration_in_minutes=DEFAULT_LESSON_DURATION_MIN,
+                    monday=group_candidate.monday,
+                    tuesday=group_candidate.tuesday,
+                    wednesday=group_candidate.wednesday,
+                    thursday=group_candidate.thursday,
+                    friday=group_candidate.friday,
+                    saturday=group_candidate.saturday,
+                    sunday=group_candidate.sunday,
+                )
+
+                group_creation_timestamp = timezone.now()
+                StatusSetter.set_status(
+                    obj=group, status=GroupStatus.PENDING, status_since=group_creation_timestamp
+                )
+                group.save()
```

Comment:
That was actually necessary to add many to many connections for teachers and students further on. Otherwise the db transaction was failing because the group id is not saved yet and it couldn't be referred to as a foreign key.

Reply:
Oh I see.